### PR TITLE
Generalize effective noise channels

### LINKED ISSUE #428: Generalize effective noise channels
So far we have added a dephasing channel for noisy `Simulation`. We can now generalize it to a depolarizing channel and more generally to any custom type of "collapse" operators for the Lindblad equation solver. Please look at QuTiP documentation [here](https://qutip.org/docs/latest/guide/dynamics/dynamics-master.html).

### PR DESCRIPTION
**Pull request by Varda-star**
Fixes #428

### CODE REVIEW

File: pulser-simulation/pulser_simulation/simconfig.py [hunk 1]
```diff
@@ -54,6 +56,11 @@ class SimConfig:
             noise types:
 
             - "dephasing": Random phase (Z) flip
+            - "depolarizing": model of decohering where the qubit undergoes a
+              bit-flip error, phase-flip error or both errors.
+            - "gen_noise": general model of noise channels where the qubit
+            undergoes a modification according to the specified noise
+            operators, it encapsulates dephasing and depolarizing channels.
```
Comment (HGSilveri) [inline pulser-simulation/pulser_simulation/simconfig.py:63]: From this description, it's not clear to me: what's the difference between having `gen_noise` vs `dephasing`+ `depolarizing`?

File: pulser-simulation/pulser_simulation/simulation.py [hunk 2]
```diff
@@ -202,7 +209,7 @@ def set_config(self, cfg: SimConfig) -> None:
                 )
             # Probability of phase (Z) flip:
             # First order in prob
-            prob = self.config.dephasing_prob / 2
+            prob = self.config.dephasing_prob
```
Comment (HGSilveri) [inline pulser-simulation/pulser_simulation/simulation.py:212]: Why this change? Isn't this still a first-order approximation? Nothing else seems to have changed.

File: pulser-simulation/pulser_simulation/simulation.py [hunk 3]
```diff
@@ -225,6 +232,97 @@ def set_config(self, cfg: SimConfig) -> None:
...
+
+            prob = self.config.depolarizing_prob
+            n = self._size
+            if prob > 0.1 and n > 1:
+                warnings.warn(
+                    "The depolarizing model is a first-order approximation"
+                    f" in the depolarizing probability. p = {2*prob}"
+                    " is too large for realistic results.",
+                    stacklevel=2,
+                )
```
Comment (HGSilveri) [inline pulser-simulation/pulser_simulation/simulation.py:254]: Shouldn't the probability be divided by 2, then?

**Review by HGSilveri — COMMENTED** [review]
Hum, some things are not so clear to me

File: pulser-simulation/pulser_simulation/simconfig.py (hunk 1 above)
Comment (Varda-star) [reply pulser-simulation/pulser_simulation/simconfig.py:63]: The difference is that you can design any operator noises.
For example: dephasing and depolarizing channels are instances of the general noise channel

File: pulser-simulation/pulser_simulation/simulation.py (hunk 2 above)
Comment (Varda-star) [reply pulser-simulation/pulser_simulation/simulation.py:212]: There is no reason to divide by two, the default probability is already less than 0.1, so I removed it.

Comment (HGSilveri) [reply pulser-simulation/pulser_simulation/simulation.py:212]: Are you sure that's the reason why it was being divided by two? It seems to me that it was rather because we are using a first order approximation (which assumes p is small) and the first term in the expansion has `p/2`. This is what the code below seems to imply.

Comment (HGSilveri) [reply pulser-simulation/pulser_simulation/simulation.py:212]: This was added in PR #233. 
`p/2` comes from the dephasing formula, actually - see eq. (2.3) in [this paper](https://arxiv.org/pdf/1509.02921.pdf).

File: pulser-simulation/pulser_simulation/simulation.py [hunk 4]
```diff
@@ -225,6 +232,97 @@ def set_config(self, cfg: SimConfig) -> None:
...
+            ]
+            self._collapse_ops += [
+                1j
+                * k
+                * (
+                    self.build_operator([("sigma_gr", [qid])]).dag()
+                    - self.build_operator([("sigma_gr", [qid])])
+                )
+                for qid in self._qid_index
+            ]
```
Comment (HGSilveri) [inline pulser-simulation/pulser_simulation/simulation.py:285]: Where did you get these Kraus operators from? They don't seem right to me

Comment (HGSilveri) [reply pulser-simulation/pulser_simulation/simulation.py:285]: Or rather, they make sense if you make `p -> p*3/4`, I think you are missing that

Comment (Varda-star) [reply pulser-simulation/pulser_simulation/simulation.py:285]: image

Comment (Varda-star) [reply pulser-simulation/pulser_simulation/simulation.py:285]: From chapter 8 on a MIT lecture on radiation.
Link: https://ocw.mit.edu/courses/22-51-quantum-theory-of-radiation-interactions-fall-2012/resources/mit22_51f12_ch8/

Comment (Varda-star) [reply pulser-simulation/pulser_simulation/simulation.py:285]: I tried then to add, M1, M2, and M3 according the uncorrelated noise approximation for many-body systems, i.e. assuming one bit is modified at the time.

Comment (HGSilveri) [reply pulser-simulation/pulser_simulation/simulation.py:285]: Ok, it seems that we have contradicting sources. Yours matches what is in Nielsen and Chuang, so I think we should stick to it

Comment (HGSilveri) [reply pulser-simulation/pulser_simulation/simulation.py:285]: But wait, wikipedia says otherwise... https://en.wikipedia.org/wiki/Quantum_depolarizing_channel

Comment (HGSilveri) [reply pulser-simulation/pulser_simulation/simulation.py:285]: Ok, I think the issue is the definition. My sources define "depolarization" as _"an operation that replaces the input state with a completely mixedstate with probability p"_, while yours state that _"with probability p an “error” occurs.
The error can be of any one of three types, where each type of error is equally likely."_ 

These are different things, so it's natural that they have different formulas.

Comment (Varda-star) [reply pulser-simulation/pulser_simulation/simulation.py:285]: I browse the Nielsen and Chuang book on Quantum COmputation and Quantum Information and they have a slight different set of operators.

Comment (HGSilveri) [reply pulser-simulation/pulser_simulation/simulation.py:285]: And look what they say right after that: 
image

This is essentially our problem here. I think we just have to pick one and clearly state what we mean by "depolarizing" in the docstring.

Comment (Varda-star) [reply pulser-simulation/pulser_simulation/simulation.py:285]: Defining the "depolarization" as "an operation that replaces the input state with a completely mixed state with probability p" is not wrong. We can start from that definition and end up with the Krauss operators I sent first.

Comment (HGSilveri) [reply pulser-simulation/pulser_simulation/simulation.py:285]: Let's see what @sebgrijalva thinks

Comment (Varda-star) [reply pulser-simulation/pulser_simulation/simulation.py:285]: @sebgrijalva told me he prefers the equation (8.102) which is in accordance with the dephasing model already chosen.

File: pulser-simulation/pulser_simulation/simulation.py [hunk 5]
```diff
@@ -225,6 +232,97 @@ def set_config(self, cfg: SimConfig) -> None:
...
+            for i in range(1, m):
+                k = np.sqrt(
+                    self.config.gen_noise_probs[i] * prob_id ** (n - 1)
+                )
+                c_op = k * self.config.gen_noise_opers[i]
+                self._collapse_ops += [
+                    self.build_operator([(c_op, [qid])])
+                    for qid in self._qid_index
+                ]
```
Comment (HGSilveri) [inline pulser-simulation/pulser_simulation/simulation.py:324]: I think you could simplify the code by making the `dephasing` and `depolarizing` cases define the proper Kraus operators and then they would always be built here. 
The current implementation introduces a lot of unnecessary duplication.

**Review by HGSilveri — CHANGES_REQUESTED** [review]

File: pulser-simulation/pulser_simulation/simconfig.py [hunk 6]
```diff
@@ -86,6 +92,13 @@ class SimConfig:
     epsilon: float = 0.01
     epsilon_prime: float = 0.05
     dephasing_prob: float = 0.05
+    depolarizing_prob: float = 0.05
+    gen_noise_probs: list[float] = field(
+        init=True, default_factory=list, repr=False
+    )
+    gen_noise_opers: list[qutip.Qobj] = field(
+        init=True, default_factory=list, repr=False
```
[the remark this answers is not in the document]
Comment (Varda-star) [reply pulser-simulation/pulser_simulation/simconfig.py]: I got an error when I set it `init = False`

[the remark this answers is not in the document]
Comment (Varda-star) [reply pulser-simulation/pulser_simulation/simconfig.py]: I removed the init arg and now it works well.

[the remark this answers is not in the document]
Comment (HGSilveri) [reply pulser-simulation/pulser_simulation/simconfig.py]: What I mean is that you don't need to specifiy it because the default is `True`

File: pulser-simulation/pulser_simulation/simulation.py [hunk 7]
```diff
@@ -225,6 +232,97 @@ def set_config(self, cfg: SimConfig) -> None:
...
+                    + self.build_operator([("sigma_gr", [qid])])
+                )
+                for qid in self._qid_index
+            ]
+            self._collapse_ops += [
+                1j
+                * k
+                * (
+                    self.build_operator([("sigma_gr", [qid])]).dag()
+                    - self.build_operator([("sigma_gr", [qid])])
```
[the remark this answers is not in the document]
Comment (Varda-star) [reply pulser-simulation/pulser_simulation/simulation.py:282]: You're right

File: pulser-simulation/pulser_simulation/simconfig.py [hunk 8]
```diff
@@ -176,3 +205,62 @@ def _calc_sigma_doppler(self) -> None:
...
+                if not isinstance(prob, float):
+                    raise TypeError(
+                        "gen_noise_probs is a list of floats"
+                        f" it must not contain a {type(prob)}."
+                    )
+
+            prob_distr = np.array(self.gen_noise_probs)
+            lower_bound = np.any(prob_distr < 0.0)
+            upper_bound = np.any(prob_distr > 1.0)
+            sum_p = sum(prob_distr) != 1.0
```
[the remark this answers is not in the document]
Comment (Varda-star) [reply pulser-simulation/pulser_simulation/simconfig.py:228]: Comparing to 0.0 will create numerical errors?

File: pulser-simulation/pulser_simulation/simconfig.py [hunk 9]
```diff
@@ -54,6 +56,10 @@ class SimConfig:
             noise types:
 
             - "dephasing": Random phase (Z) flip
+            - "depolarizing": model of decohering where the qubit undergoes a
+              bit-flip error, phase-flip error or both errors.
```
[the remark this answers is not in the document]
Comment (Varda-star) [reply pulser-simulation/pulser_simulation/simconfig.py]: Yes as it is in the book
image

[the remark this answers is not in the document]
Comment (Varda-star) [reply pulser-simulation/pulser_simulation/simconfig.py]: Then from this we derive the equation (8.102)
image

[the remark this answers is not in the document]
Comment (Varda-star) [reply pulser-simulation/pulser_simulation/simconfig.py]: You're right

[the remark this answers is not in the document]
Comment (Varda-star) [reply pulser-simulation/pulser_simulation/simconfig.py]: I will try something else.

File: pulser-simulation/pulser_simulation/simconfig.py [hunk 10]
```diff
@@ -54,6 +56,10 @@ class SimConfig:
...
+            - "gen_noise": general noise channel through which we can custom
```
[the remark this answers is not in the document]
Comment (Varda-star) [reply pulser-simulation/pulser_simulation/simconfig.py]: You mean the way I wrote the code only allow on of the three to exist?

File: pulser-simulation/pulser_simulation/simconfig.py [hunk 11]
```diff
@@ -86,6 +92,13 @@ class SimConfig:
     epsilon: float = 0.01
     epsilon_prime: float = 0.05
     dephasing_prob: float = 0.05
+    depolarizing_prob: float = 0.05
+    gen_noise_probs: list[float] = field(
+        init=True, default_factory=list, repr=False
+    )
```
[the remark this answers is not in the document]
Comment (Varda-star) [reply pulser-simulation/pulser_simulation/simconfig.py]: It's because of the approximation we use for many bodies. 
All effective channels are not probability-wise as the amplitude damping and the phase damping channel.

[the remark this answers is not in the document]
Comment (Varda-star) [reply pulser-simulation/pulser_simulation/simconfig.py]: Our noise model is based on the fact that only one bit string is modified at the time, and it is only justified when "rates" of collapse operators are low.

[the remark this answers is not in the document]
Comment (Varda-star) [reply pulser-simulation/pulser_simulation/simconfig.py]: This is why I wanted to seperate collapse operators from their rates to make sure the approx is verified. Normally I must not allow any set of operators where all rates aren't all low. It will not give realistic result. Or as done for the dephasing and depolarizing channel, warn the user that the effective channel he proposed will not yield realistic results since all rates are not low enough.

File: pulser-simulation/pulser_simulation/simconfig.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (HGSilveri) [reply pulser-simulation/pulser_simulation/simconfig.py]: Yes, you are overriding the `_colase_ops` on each case

File: pulser-simulation/pulser_simulation/simconfig.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (Varda-star) [reply pulser-simulation/pulser_simulation/simconfig.py]: And as it's done for the dephasing and depolarizing channels, I warn the user whenever the rate of an operator is not low enough that it would yield unrealistic results according to the approx we used.

File: pulser-simulation/pulser_simulation/simulation.py (hunk 5 above)
Comment (Varda-star) [reply pulser-simulation/pulser_simulation/simulation.py:324]: I must remove them from the set of noise models?

Comment (HGSilveri) [reply pulser-simulation/pulser_simulation/simulation.py:324]: No no, they are convenient to have. But you could store their noise operators under `gen_noise_opers` so that the collapse operators are built always in the same place

File: pulser-simulation/pulser_simulation/simconfig.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (Varda-star) [reply pulser-simulation/pulser_simulation/simconfig.py]: How can I?

File: pulser-simulation/pulser_simulation/simconfig.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (Varda-star) [reply pulser-simulation/pulser_simulation/simconfig.py]: So if I put a `+=` instead of `=` it's solved?

File: pulser-simulation/pulser_simulation/simulation.py (hunk 5 above)
Comment (Varda-star) [reply pulser-simulation/pulser_simulation/simulation.py:324]: Because the only place where collapse_operators are defined is in `dephasing`, `depolarizing` or `gen_noise` mode

Comment (Varda-star) [reply pulser-simulation/pulser_simulation/simulation.py:324]: I think `effective_nosie` is more convenient than `gen_noise`. What do you think?

File: pulser-simulation/pulser_simulation/simconfig.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (HGSilveri) [reply pulser-simulation/pulser_simulation/simconfig.py]: No, you can just add more and more Kraus operators. It's tricky, to be honest. I think to start the best thing is to allow only one of the three. If the user wants to combine multiple noises, they still can do so with the general noise operators option

File: pulser-simulation/pulser_simulation/simulation.py (hunk 5 above)
Comment (HGSilveri) [reply pulser-simulation/pulser_simulation/simulation.py:324]: I agree, I prefer `eff_noise` over `gen_noise` too

File: pulser-simulation/pulser_simulation/simconfig.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (HGSilveri) [reply pulser-simulation/pulser_simulation/simconfig.py]: Well, the probability associated with each operator (if the operators already include the probablity, as I'm suggesting) will simply be `op.dag() * op` right?

File: pulser-simulation/pulser_simulation/simulation.py (hunk 5 above)
Comment (Varda-star) [reply pulser-simulation/pulser_simulation/simulation.py:324]: No no, they are convenient to have. But you could store their noise operators under gen_noise_opers so that the collapse operators are built always in the same place// If I do so it will create confusion since I iterate it in a loop to build Kraus operators.

File: pulser-simulation/pulser_simulation/simconfig.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (Varda-star) [reply pulser-simulation/pulser_simulation/simconfig.py]: Ah ok, so `dephasing + depolarizing` will be prohibited?

File: pulser-simulation/pulser_simulation/simconfig.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (HGSilveri) [reply pulser-simulation/pulser_simulation/simconfig.py]: If the probability is incorporated in the operator, it wouldn't be just `sigma_x`, it would be `op = sqrt(p)*sigma_x`

File: pulser-simulation/pulser_simulation/simconfig.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (HGSilveri) [reply pulser-simulation/pulser_simulation/simconfig.py]: Yes, for now we can allow only one of `dephasing`, `depolarizing` or `gen_noise`, but never more than one at a time

File: pulser-simulation/pulser_simulation/simulation.py (hunk 5 above)
Comment (HGSilveri) [reply pulser-simulation/pulser_simulation/simulation.py:324]: I think you are missing the point. In all three cases, all you are doing is taking the probabilities, building the single-qubit Kraus operators and then building the collapse operators from them. The difference in each case is contained in the single-qubit Kraus operators, so the collapse operators could be built from them in one place rather than repeated every time

File: pulser-simulation/pulser_simulation/simconfig.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (Varda-star) [reply pulser-simulation/pulser_simulation/simconfig.py]: So it is hard to extract p. In the case of pauli matrices it's simple but for a general Kraus Operator it isn't possible. The only relation we know about them is that `sum(op_k) = id`

[the remark this answers is not in the document]
Comment (HGSilveri) [reply pulser-simulation/pulser_simulation/simconfig.py]: Yes, I think you are right. I was assuming the operator was a unitary, but this is not a requirement. Nevermind my suggestion then.

File: pulser-simulation/pulser_simulation/simconfig.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (Varda-star) [reply pulser-simulation/pulser_simulation/simconfig.py]: What do you think about adding this restriction in the method `_check_noise_types` in simconfig?

[the remark this answers is not in the document]
Comment (HGSilveri) [reply pulser-simulation/pulser_simulation/simconfig.py]: I think it's a good idea, it's the right place for that check

File: pulser-simulation/pulser_simulation/simulation.py (hunk 5 above)
Comment (HGSilveri) [reply pulser-simulation/pulser_simulation/simulation.py:324]: And why not?

Comment (Varda-star) [reply pulser-simulation/pulser_simulation/simulation.py:324]: In either `dephasing` , `depolarizing` or `eff_noise`, i will build Kraus operators then build all collapse operators at once.

Comment (HGSilveri) [reply pulser-simulation/pulser_simulation/simulation.py:324]: Exactly, don't you think it's cleaner this way?

Comment (Varda-star) [reply pulser-simulation/pulser_simulation/simulation.py:324]: Since Kraus operators are built only in one of these modes, I need that storage variable to be defined beforehand, why not defining it as an attribute of the class?

Comment (Varda-star) [reply pulser-simulation/pulser_simulation/simulation.py:324]: `self._Kraus_ops`

Comment (HGSilveri) [reply pulser-simulation/pulser_simulation/simulation.py:324]: Can't you do it before the first `if`?

Comment (HGSilveri) [reply pulser-simulation/pulser_simulation/simulation.py:324]: I think an attribute is not necessary if you build the collapse operators in the same call

Comment (Varda-star) [reply pulser-simulation/pulser_simulation/simulation.py:324]: Ok, I will do it before the first `if`.

File: pulser-simulation/pulser_simulation/simconfig.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (Varda-star) [reply pulser-simulation/pulser_simulation/simconfig.py]: Let's cook some code then

File: tests/test_simulation.py [hunk 12]
```diff
@@ -590,19 +613,17 @@ def test_dephasing():
...
+    pulse = Pulse.ConstantPulse(duration, np.pi, 0, 0)
     seq.add(pulse, "ch0")
     sim = Simulation(
         seq, sampling_rate=0.01, config=SimConfig(noise="dephasing")
     )
     assert sim.run().sample_final_state() == Counter({"0": 595, "1": 405})
     assert len(sim._collapse_ops) != 0
     with pytest.warns(UserWarning, match="first-order"):
-        reg = Register.from_coordinates([(0, 0), (0, 10)], prefix="q")
+        reg = Register.from_coordinates([(0, 0), (0, 7)], prefix="q")
```
Comment (HGSilveri) [inline tests/test_simulation.py:624]: Why this change?

File: tests/test_simulation.py [hunk 13]
```diff
@@ -611,7 +632,67 @@ def test_dephasing():
...
+    reg = Register.from_coordinates([(0, 0)], prefix="q")
+    seq = Sequence(reg, Chadoq2)
+    seq.declare_channel("ch0", "rydberg_global")
+    duration = 2500
+    pulse = Pulse.ConstantPulse(duration, np.pi, 0, 0)
+    seq.add(pulse, "ch0")
+    sim = Simulation(
+        seq, sampling_rate=0.01, config=SimConfig(noise="depolarizing")
+    )
+    assert sim.run().sample_final_state() == Counter({"0": 587, "1": 413})
```
Comment (HGSilveri) [inline tests/test_simulation.py:646]: Check also that the state is not pure anymore, otherwise you can't distinguish between a mixed state and an equal superpostion

File: tests/test_simulation.py [hunk 14]
```diff
@@ -611,7 +632,67 @@ def test_dephasing():
...
+    seq.add(pulse, "ch0")
+    sim = Simulation(
+        seq,
+        sampling_rate=0.01,
+        config=SimConfig(
+            noise="eff_noise",
+            eff_noise_opers=[matrices["I"], matrices["Z"]],
+            eff_noise_probs=[0.95, 0.05],
+        ),
+    )
```
Comment (HGSilveri) [inline tests/test_simulation.py:676]: If I'm not mistaken, this should be the same as dephasing with `p=0.10`. Why don't you run a simulation with that config and check that the final states match?

**Review by HGSilveri — COMMENTED** [review]
Much better!

File: tests/test_simulation.py (hunk 12 above)
Comment (Varda-star) [reply tests/test_simulation.py:624]: No real reason in reality

File: tests/test_simulation.py (hunk 13 above)
Comment (Varda-star) [reply tests/test_simulation.py:646]: I took the trace of the square of the density matrix at the end of the evolution and checked if it's lower than one without being close to one.

Comment (HGSilveri) [reply tests/test_simulation.py:646]: Correct me if I'm wrong, but it should actually be close to `1-3p/4`, no? Perhaps you can check for that

Comment (Varda-star) [reply tests/test_simulation.py:646]: Is it enough? Because I made the same test for the dephasing channel and the test below worked. Does it mean the state is a mixed state?

Comment (Varda-star) [reply tests/test_simulation.py:646]: The absolute difference is 0.03, is it close enough?

**Comment by HGSilveri** [comment]
The docs are failing because of an update to the `sphinx_autodoc_typehints` extension. Would be best to pin it to the previous release (ie `sphinx_autodoc_typehints == 1.21.3`) in `docs/requirements.txt`

File: tests/test_simulation.py (hunk 13 above)
Comment (HGSilveri) [reply tests/test_simulation.py:646]: Are you struggling with this?

Comment (Varda-star) [reply tests/test_simulation.py:646]: Yes. I don't understand why `1- 3*P`. With `rho = p/2 I + (1-p) 1` I found `tr(rho**2) = 2 (1-3/4p) ** 2 + 3/8 p**2`

Comment (Varda-star) [reply tests/test_simulation.py:646]: By expanding `rho**2` and using `tr(sigma_i) = 0` and `tr(sigma_i * sigma_j) = 0` for `i!=j`

File: tests/test_simconfig.py [hunk 15]
```diff
@@ -70,26 +70,23 @@ def test_init(matrices):
...
-        NotImplementedError,
-        match="Depolarizing, dephasing and eff_noise channels",
-    ):
-        SimConfig(noise=("dephasing", "eff_noise"))
-    with pytest.raises(
-        NotImplementedError,
-        match="Depolarizing, dephasing and eff_noise channels",
-    ):
-        SimConfig(noise=("dephasing", "depolarizing", "eff_noise"))
+        SimConfig(noise=noise_sample)
```
Comment (HGSilveri) [inline tests/test_simconfig.py:89]: These should be isolated in a single test, otherwise you are repeating the rest of the function needlessly

File: tests/test_simconfig.py [hunk 16]
```diff
@@ -70,26 +70,23 @@ def test_init(matrices):
...
+@pytest.mark.parametrize(
+    "noise_sample,",
+    [
+        ("dephasing", "depolarizing"),
+        ("eff_noise", "depolarizing"),
+        ("eff_noise", "dephasing"),
+        ("depolarizing", "eff_noise", "dephasing"),
+    ],
+)
+def test_init2(matrices, noise_sample):
```
Comment (HGSilveri) [inline tests/test_simconfig.py:84]: Give a more descriptive name, like `test_multiple_kraus_noise`

File: tests/test_simulation.py (hunk 12 above)
Comment (HGSilveri) [reply tests/test_simulation.py:624]: Then you should revert it

File: tests/test_simconfig.py [hunk 17]
```diff
@@ -40,3 +70,79 @@ def test_init():
...
+            noise=("eff_noise"),
+            eff_noise_opers=[matrices["I"], matrices["X"]],
+            eff_noise_probs=[0.5, 2.0],
+        )
+    with pytest.raises(ValueError, match="is not a probability distribution."):
+        SimConfig(
+            noise=("eff_noise"),
+            eff_noise_opers=[matrices["I"], matrices["X"]],
+            eff_noise_probs=[0.3, 0.2],
+        )
```
[the remark this answers is not in the document]
Comment (HGSilveri) [reply tests/test_simconfig.py:115]: You didn't separate these in the end

File: pulser-simulation/pulser_simulation/simconfig.py [hunk 18]
```diff
@@ -206,29 +213,36 @@ def _calc_sigma_doppler(self) -> None:
...
-            if self.gen_noise_opers == [] or self.gen_noise_probs == []:
+        if "eff_noise" in self.noise:
+            if len(self.eff_noise_opers) != len(self.eff_noise_probs):
+                raise ValueError(
+                    f"The operators list length({len(self.eff_noise_opers)}) "
+                    "and probabilities list length"
+                    f"({len(self.eff_noise_probs)}) must be equal."
+                )
+            if self.eff_noise_opers == [] or self.eff_noise_probs == []:
                 raise ValueError("Fill the general noise parameters.")
```
[the remark this answers is not in the document]
Comment (HGSilveri) [reply pulser-simulation/pulser_simulation/simconfig.py:222]: You didn't change this

File: pulser-simulation/pulser_simulation/simulation.py [hunk 19]
```diff
@@ -253,75 +248,56 @@ def set_config(self, cfg: SimConfig) -> None:
...
                 k = np.sqrt(
-                    self.config.gen_noise_probs[i] * prob_id ** (n - 1)
+                    self.config.eff_noise_probs[i] * prob_id ** (n - 1)
                 )
-                c_op = k * self.config.gen_noise_opers[i]
-                self._collapse_ops += [
-                    self.build_operator([(c_op, [qid])])
-                    for qid in self._qid_index
-                ]
+                K_op = k * self.config.eff_noise_opers[i]
```
[the remark this answers is not in the document]
Comment (HGSilveri) [reply pulser-simulation/pulser_simulation/simulation.py:292]: Didn't change this either

File: tests/test_simconfig.py [hunk 20]
```diff
@@ -40,3 +70,84 @@ def test_init():
...
+        match="Depolarizing, dephasing and eff_noise channels",
+    ):
+        SimConfig(noise=noise_sample)
+
+
+@pytest.mark.parametrize(
+    "oper_list",
+    [
+        [qeye(2), sigmax()],
+    ],
```
Comment (HGSilveri) [inline tests/test_simconfig.py:96]: If `oper_list` is always the same, there's no reason for it to be a variable

File: tests/test_simconfig.py [hunk 21]
```diff
@@ -40,3 +70,84 @@ def test_init():
...
+    with pytest.raises(ValueError, match="is not a probability distribution."):
+        SimConfig(
+            noise=("eff_noise"),
+            eff_noise_opers=oper_list,
+            eff_noise_probs=prob_distr,
+        )
+    with pytest.raises(TypeError, match="eff_noise_probs is a list of floats"):
+        SimConfig(
+            noise=("eff_noise"), eff_noise_probs=[""], eff_noise_opers=[""]
+        )
```
Comment (HGSilveri) [inline tests/test_simconfig.py:116]: This should be in `test_eff_noise_opers`, I guess, just so it's not repeated

**Review by HGSilveri — APPROVED** [review]
LGTM!
